### PR TITLE
oop-static_modifier-hard-q4

### PR DESCRIPTION
**Pull request by KaiKhademi**
## PR Checklist
- [x]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [x]  Are the assertion fail messages good?
    - [x]  Logical (does it make sense?)
    - [x]  Readable (can a student understand what’s wrong?)
    - [x]  Not too feed-y (does it give away the answer?)
    - [x]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [x]  Specific (does the fail message relate to the question?)
- [x]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [x]  EOF newlines
- [x]  Do the comments for Parsons follow the correct convention?
- [x]  Is the question being tested sufficiently?
    - [x]  Are there representative cases to test the program?
    - [x]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [x]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”

### CODE REVIEW

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]
Please see my feedback on other questions in this category and apply it to this PR.

File: src/oop/static_modifier/hard/q4/MainTest.java [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    public void bedHasRequiredConstructor() {
+        String missingConstructorMessage = "The %s class is missing a required constructor. Make sure that it is named correctly and has the correct parameters.";
+        String wrongAccessModifier = "The %s class constructor has the wrong visibility modifier. Make sure that it is visible from all other classes.";
+        Class<?>[] constructorArgs = new Class[]{int.class};
+        assertTrue(testClass.hasConstructor(constructorArgs), String.format(missingConstructorMessage, className));
+        assertTrue(testClass.hasModifier(constructorArgs, "public"), String.format(wrongAccessModifier, className));
+    }
+
+    @Test
+    @Order(3)
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q4/MainTest.java:67]: order this as the first test

File: src/oop/static_modifier/hard/q4/MainTest.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+        Object[][] constructorArgs = {
+                {value1, int.class}
+        };
+        Object checkupInstance = testClass.createInstance(constructorArgs);
+        _assertEquals(value1, testClass.getFieldValue(checkupInstance, attributeName), String.format(wrongValueMessage, className, attributeName));
+    }
+
+    @Test
+    @Order(5)
+    public void makeFurnitureIsDefinedCorrectly() {
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q4/MainTest.java:104]: this test is no longer needed

File: src/oop/static_modifier/hard/q4/MainTest.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    @Tag("dependent1")
+    @Order(7)
+    public void correctMainMethod() throws Throwable {
+        Clause[] clauses = {
+                new StringLiteral("4 pieces of furniture were made"),
+                new Optional(new StringLiteral(" "))
+        };
+        Object classInstance = classObject.createInstance();
+        classObject.callMethod("main", new Object[][]{{new String[0], String[].class}}, new String[]{"public"}, classInstance,
+                clauses);
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q4/MainTest.java:154]: add an error message

File: src/oop/static_modifier/hard/q4/MainTest.java [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        }
+        testClass.setFieldValue(null, initialBeams, staticAttributeName);
+    }
+
+    @Test
+    @Tag("dependent1")
+    @Order(7)
+    public void correctMainMethod() throws Throwable {
+        Clause[] clauses = {
+                new StringLiteral("4 pieces of furniture were made"),
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q4/MainTest.java:149]: make this a variable at the top of the tests

**Review by SamamthaH — CHANGES_REQUESTED** [review]

File: src/oop/static_modifier/hard/q4/QuestionText.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+      visible from other classes.
+    - An `int` called **numBeams** which is the amount of wood required to make the `Bed` Object. Note that it should not be
+      visible from other classes.
+
+- Constructors:
+    - A constructor that is accessible from other methods, and accepts an `int` as a parameter to set the **numBeams**
+      attribute of the object.
+
+- Methods:
+    - Note that all methods in this class should be visible from another class
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q4/QuestionText.md:27]: "all methods" doesn't make sense when there is only one method

File: src/oop/static_modifier/hard/q4/QuestionText.md [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    - A `makeFurniture` method which return false if the **numBeams** value is more than **woodBeams**, meaning that the
+      `Bed` object cannot be made, or subtract the **numBeams** value from the **woodBeams** and return true.
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q4/QuestionText.md:29]:
```suggestion
    - A `makeFurniture` method which returns false if the **numBeams** value is more than **woodBeams**, meaning that the
      `Bed` object cannot be made, or subtracts the **numBeams** value from the **woodBeams** and returns true.
```

File: src/oop/static_modifier/hard/q4/MainTest.java [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+        assertTrue(testClass.hasConstructor(constructorArgs), String.format(missingConstructorMessage, className));
+        assertTrue(testClass.hasModifier(constructorArgs, "public"), String.format(wrongAccessModifier, className));
+    }
+
+    @Test
+    @Order(1)
+    @Tag("dependency1")
+    public void bedIsInitializedCorrectly() {
+        _assertEquals(initialBeams, testClass.getFieldValue(null, staticAttributeName),
+                "Your " + staticAttributeName + " static attribute is not initialized correctly.");
```
Comment (SamamthaH) [inline src/oop/static_modifier/hard/q4/MainTest.java:72]: check the attribute exists before checking its value

**Review by SamamthaH — CHANGES_REQUESTED** [review]

**Pull request merged**